rust/src/jitter.rs

```diff
@@ -0,0 +1,368 @@
+use std::{str::FromStr, sync::Arc};
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use drift::state::user::{OrderStatus, User, UserStats};
+use drift_sdk::{
+    auction_subscriber::{AuctionSubscriber, AuctionSubscriberConfig}, 
+    constants::PROGRAM_ID as drift_program, 
+    event_emitter::Event,
+    types::{
+        CommitmentConfig, 
+        MarketType, 
+        Order, 
+        PerpMarket, ReferrerInfo
+    }, websocket_program_account_subscriber::ProgramAccountUpdate, AccountProvider, DriftClient, Pubkey, Wallet
+};
+use jit_proxy::state::PriceType;
+use solana_sdk::signature::Signature;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::task::JoinHandle;
+
+use crate::jit_proxy_client::{JitIxParams, JitProxyClient};
+use crate::JitResult;
+
+
+type UserFilter = dyn Fn(&User, &String, Order) -> bool + Send + Sync;
+
+fn log_details(order: &Order) {
+    log::info!("Market Type: {:?}", order.market_type);  
+    log::info!("Market index: {}", order.market_index);
+    log::info!("Order price: {}", order.price);
+    log::info!("Order direction: {:?}", order.direction);
+    log::info!("Auction start price: {}", order.auction_start_price);
+    log::info!("Auction end price: {}", order.auction_end_price); 
+    log::info!("Auction duration: {} slots", order.auction_duration);
+    log::info!("Order base asset amount: {}", order.base_asset_amount);
+    log::info!("Order base asset amount filled: {}", order.base_asset_amount_filled);
+}
+
+#[inline(always)]
+fn check_err(err: String, order_sig: String) -> Option<()> {
+    if err.contains("0x1770") || err.contains("0x1771") {
+        log::error!("Order: {} does not cross params yet, retrying", order_sig);
+        None
+    } else if err.contains("0x1779") {
+        log::error!("Order: {} could not fill, retrying", order_sig);
+        None
+    } else if err.contains("0x1793") {
+        log::error!("Oracle invalid, retrying");
+        None
+    } else if err.contains("0x1772") {
+        log::error!("Order: {} already filled", order_sig);
+        Some(())
+    } else {
+        log::error!("Error: {}", err);
```

Comment:
might be good to include order_sig here for reference (same with the oracle_invalid case